pkg/sentry/fsimpl/fuse/dev_test.go

```diff
@@ -0,0 +1,373 @@
+// Copyright 2020 The gVisor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fuse
+
+import (
+	"io"
+	"testing"
+
+	"gvisor.dev/gvisor/pkg/abi/linux"
+	"gvisor.dev/gvisor/pkg/sentry/fsimpl/testutil"
+	"gvisor.dev/gvisor/pkg/sentry/fsimpl/tmpfs"
+	"gvisor.dev/gvisor/pkg/sentry/kernel"
+	"gvisor.dev/gvisor/pkg/sentry/kernel/auth"
+	"gvisor.dev/gvisor/pkg/sentry/vfs"
+	"gvisor.dev/gvisor/pkg/sync"
+	"gvisor.dev/gvisor/pkg/usermem"
+	"gvisor.dev/gvisor/tools/go_marshal/marshal"
+)
+
+const testOpcode linux.FUSEOpcode = 1000
+
+var testReadMu sync.Mutex
+
+type testObject struct {
+	opcode linux.FUSEOpcode
+}
+
+// TestFUSECommunication tests that the communication layer between the Sentry and the
+// FUSE server daemon works as expected.
+func TestFUSECommunication(t *testing.T) {
+	s := setup(t)
+	defer s.Destroy()
+
+	k := kernel.KernelFromContext(s.Ctx)
+	creds := auth.CredentialsFromContext(s.Ctx)
+
+	fuseConn, fd, err := newTestConnection(s, k)
+	if err != nil {
+		t.Fatalf("newTestConnection: %v", err)
+	}
+
+	// Create test cases with different number of concurrent clients and servers.
+	// A server in this context is a task/thread that serves upto one request.
+	// And so there must be atleast as many servers as clients.
+	testCases := []struct {
+		Name       string
+		NumClients int
+		NumServers int
+	}{
+		{
+			Name:       "SingleClientSingleServer",
+			NumClients: 1,
+			NumServers: 1,
+		},
+		{
+			Name:       "SingleClientMultipleServers",
+			NumClients: 1,
+			NumServers: 10,
+		},
+		{
+			Name:       "MultipleClientsMultipleServers",
+			NumClients: 10,
+			NumServers: 10,
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.Name, func(t *testing.T) {
+			if testCase.NumServers < testCase.NumClients {
+				t.Fatalf("Need atleast %v servers, but have %v", testCase.NumClients, testCase.NumServers)
```

Comment:
This looks like a bug in the test code, not a bug of the code? Failing one test case due to a bug in the test slightly makes no sense. Proposal: at line 80 (right after the for loop), add this check and `continue` and `Warn` if bad test given

Reply:
Made some changes to the test code to make all this simpler. The server will now behave like an actual server.